### PR TITLE
python3Packages.{mautrix,matrix-nio}: add `withOlm` flags

### PR DESCRIPTION
**Pull request by emilazy**
## Description of changes

An alternate approach to https://github.com/NixOS/nixpkgs/pull/336694. This has some nice properties: we don’t have to fish into dependency trees or fiddle around with hacks like overriding `knownVulnerabilities`, Hydra never builds libolm, but all available functionality of the library can be tested. The disabled tests and patching are what we’d have to do if we went the route of never testing Olm functionality anyway.

~~On the other hand, the definition of `olmAllowed` is kind of gross? It might be better if these packages took a configuration argument directly asking whether they’re going to be used with E2E, and set their dependencies and tests accordingly. Not sure. Putting it up here for review.~~ I did this.

cc @AndrewKvalheim

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by emilazy** [comment]
Okay, I made this more normal and just added `withOlm` flags. This is recommended against in the Python documentation because it can result in conflicting versions, but it seems to be the only approach that lets us continue to run the encryption‐related tests when the functionality is going to be exercised without weird hacks. The result looks like an okay compromise to me now and should clear up the issues. It’s possible that I missed a package expecting the E2E functionality when adding the overrides, but I tried to be thorough. I’d appreciate it if you could take a look, @AndrewKvalheim @antifuchs.

Once this is merged and backported I’ll go ahead with https://github.com/NixOS/nixpkgs/pull/335189, as the fallout so far seems to have been minimal except for the mistake with `element-desktop` and these Python libraries.

Result of `nixpkgs-review pr 336901` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>19 packages built:</summary>
  <ul>
    <li>heisenbridge</li>
    <li>heisenbridge.dist</li>
    <li>home-assistant-component-tests.matrix</li>
    <li>python311Packages.matrix-nio</li>
    <li>python311Packages.matrix-nio.dist</li>
    <li>python311Packages.mautrix (python311Packages.mautrix-appservice)</li>
    <li>python311Packages.mautrix.dist (python311Packages.mautrix-appservice.dist)</li>
    <li>python311Packages.zulip</li>
    <li>python311Packages.zulip.dist</li>
    <li>python312Packages.matrix-nio</li>
    <li>python312Packages.matrix-nio.dist</li>
    <li>python312Packages.mautrix (python312Packages.mautrix-appservice)</li>
    <li>python312Packages.mautrix.dist (python312Packages.mautrix-appservice.dist)</li>
    <li>python312Packages.zulip</li>
    <li>python312Packages.zulip.dist</li>
    <li>visidata</li>
    <li>visidata.dist</li>
    <li>zulip-term</li>
    <li>zulip-term.dist</li>
  </ul>
</details>

Result of `nixpkgs-review pr 336901 --extra-nixpkgs-config '{ allowInsecurePredicate = x: true; }'` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>36 packages built:</summary>
  <ul>
    <li>heisenbridge</li>
    <li>heisenbridge.dist</li>
    <li>home-assistant-component-tests.matrix</li>
    <li>matrix-commander</li>
    <li>matrix-commander.dist</li>
    <li>maubot (python312Packages.maubot)</li>
    <li>maubot.dist (python312Packages.maubot.dist)</li>
    <li>mautrix-facebook</li>
    <li>mautrix-facebook.dist</li>
    <li>mautrix-googlechat</li>
    <li>mautrix-googlechat.dist</li>
    <li>mautrix-telegram</li>
    <li>mautrix-telegram.dist</li>
    <li>opsdroid</li>
    <li>opsdroid.dist</li>
    <li>pantalaimon</li>
    <li>pantalaimon-headless</li>
    <li>pantalaimon-headless.dist</li>
    <li>pantalaimon.dist</li>
    <li>python311Packages.matrix-nio</li>
    <li>python311Packages.matrix-nio.dist</li>
    <li>python311Packages.mautrix (python311Packages.mautrix-appservice)</li>
    <li>python311Packages.mautrix.dist (python311Packages.mautrix-appservice.dist)</li>
    <li>python311Packages.zulip</li>
    <li>python311Packages.zulip.dist</li>
    <li>python312Packages.matrix-nio</li>
    <li>python312Packages.matrix-nio.dist</li>
    <li>python312Packages.mautrix (python312Packages.mautrix-appservice)</li>
    <li>python312Packages.mautrix.dist (python312Packages.mautrix-appservice.dist)</li>
    <li>python312Packages.zulip</li>
    <li>python312Packages.zulip.dist</li>
    <li>visidata</li>
    <li>visidata.dist</li>
    <li>weechatScripts.weechat-matrix</li>
    <li>zulip-term</li>
    <li>zulip-term.dist</li>
  </ul>
</details>

**Comment by Ma27** [comment]
It's a Python library, so is it really that much of an effort to allow-list it in one's nixpkgs config and build it yourself? Is the use-case to just build nio without the whitelisting?

I don't have a strong opinion on that as long as we don't ship any packages like mautrix-* w/o encryption by default, just trying to understand what we want to achieve :) 

> This is recommended against in the Python documentation because it can result in conflicting versions

cc @mweinelt

File: pkgs/development/python-modules/mautrix/default.nix
Comment (SuperSandro2000) [inline pkgs/development/python-modules/mautrix/default.nix:54]:
```suggestion
  optional-dependencies = {
```

File: pkgs/development/python-modules/mautrix/default.nix
Comment (SuperSandro2000) [inline pkgs/development/python-modules/mautrix/default.nix:51]: We can turn this into a real extra

https://github.com/mautrix/python/blob/v0.20.6/setup.py#L33

File: pkgs/servers/mautrix-facebook/default.nix
Comment (SuperSandro2000) [inline pkgs/servers/mautrix-facebook/default.nix:24]: and then use mautrix.optional-dependencies.encryption here

File: pkgs/development/python-modules/matrix-nio/default.nix
Comment (SuperSandro2000) [inline pkgs/development/python-modules/matrix-nio/default.nix:73]: Please keep the optional-dependencies. That's the right approach here https://github.com/matrix-nio/matrix-nio/blob/0.25.0/pyproject.toml#L37

File: pkgs/development/python-modules/matrix-nio/default.nix
Comment (SuperSandro2000) [inline pkgs/development/python-modules/matrix-nio/default.nix:80]:
```suggestion
lib.optionals withOlm optional-dependencies.e2
```

**Comment by SuperSandro2000** [comment]
> This is recommended against in the Python documentation because it can result in conflicting versions, but it seems to be the only approach that lets us continue to run the encryption‐related tests when the functionality is going to be exercised without weird hacks.

We want to do that and it should now be catched by our hooks in case we mix things.

Also test dependencies don't mix with the normal ones so that should be even less of an issue.

File: pkgs/development/python-modules/mautrix/default.nix
Comment (emilazy) [reply pkgs/development/python-modules/mautrix/default.nix:51]: Or is the idea that the `optional-dependencies` would still be there but the `withOlm` flag would gate the tests? That would mean that downstream consumers could forget to enable it and be using untested functionality, but I guess there’s unlikely to be new ones added.

**Comment by emilazy** [comment]
> It's a Python library, so is it really that much of an effort to allow-list it in one's nixpkgs config and build it yourself? Is the use-case to just build nio without the whitelisting?
> 
> I don't have a strong opinion on that as long as we don't ship any packages like mautrix-* w/o encryption by default, just trying to understand what we want to achieve :)

I’m just trying to fix https://github.com/NixOS/nixpkgs/issues/336052 and the like. You shouldn’t get a security warning for the packages that already don’t use the E2E functionality and aren’t going to exercise anything from libolm.

**Comment by mweinelt** [comment]
The optional-dependencies definition must stay.

**Comment by emilazy** [comment]
I’ve pushed a change that restores the `optional-dependencies` but should hopefully still maintain all the desired properties; please take a look. I can also just unconditionally disable the Olm‐related tests if people would prefer to not do anything at all weird here, but it makes me a little uneasy given that Hydra is not going to flag up any issues with those configurations.

**Comment by emilazy** [comment]
Bumping this as it’s the main remaining blocker to the 24.05 libolm backport.

File: pkgs/development/python-modules/mautrix/default.nix
Comment (Ma27) [inline pkgs/development/python-modules/mautrix/default.nix:48]: Isn't this supposed to be under `passthru`?

File: nixos/tests/matrix/mjolnir.nix
Comment (Ma27) [inline nixos/tests/matrix/mjolnir.nix:113]: Can't we use the old approach here of adding optional dependencies? That we we'd no longer risk having two distinct matrix-nios / mautrix packages in an env.

**Review by Ma27 — COMMENTED** [review]
> You shouldn’t get a security warning for the packages that already don’t use the E2E functionality and aren’t going to exercise anything from libolm

Fair.
My main objectives still seem to be kept (no downgrade of defaults for bridges, no silence of security warnings when using said bridges with libolm).

---

There are two things we should probably address before merging.

File: pkgs/development/python-modules/mautrix/default.nix
Comment (emilazy) [reply pkgs/development/python-modules/mautrix/default.nix:48]: Sandro [told me to change it](https://github.com/NixOS/nixpkgs/pull/336901#discussion_r1731098403). I guess this way is new?

File: nixos/tests/matrix/mjolnir.nix
Comment (emilazy) [reply nixos/tests/matrix/mjolnir.nix:113]: That would mean that the libolm functionality would be used without its tests having been run, which is the invariant I’m trying to maintain. We could just disable the encryption‐related tests unconditionally, but given that it is probably going to be quite a long time before all of these bridges are using something else, it’s pretty critical functionality, and there are a lot of those tests, that feels like it would be a regression. I’m happy to do that if people really don’t like this solution and don’t think testing the encryption functionality is important, and I’m up for considering any better way to achieve the invariant, but that’s why it’s structured the way it is currently.

File: pkgs/development/python-modules/mautrix/default.nix
Comment (Ma27) [reply pkgs/development/python-modules/mautrix/default.nix:48]: Apparently yes, had to look around a bit before I found this pattern elsewhere, sorry!

File: nixos/tests/matrix/mjolnir.nix
Comment (Ma27) [reply nixos/tests/matrix/mjolnir.nix:113]: Apologies, I missed that part above!

I agree that the additional safety-net of running the tests then is a good idea. And Sandro is right, the setup hooks should catch conflicting packages in the first place. While I still think the solution is kinda messy, I don't know a better way to do this and the situation we're in is also kind of messy.

One may argue that we could just keep these two packages being marked as insecure, but apparently most of the other folks here have a different take on this and it's not a hill worth dying on.

Resolving.

**Review by Ma27 — APPROVED** [review]
I don't particularly like it, but I'm fine if this gets merged.

Implementation-wise the change looks correct.

**Comment by emilazy** [comment]
I also don’t particularly like it :) But I can’t think of a better compromise and I don’t want more packages than necessary to be marked insecure by this, so merging.

**Pull request merged**
